### PR TITLE
[PoC][WIP, not ready for review]feat: add abstractions in authentication module

### PR DESCRIPTION
**Pull request by jannyHou**
Signed-off-by: jannyHou <juehou@ca.ibm.com>

connect to https://github.com/strongloop/loopback-next/issues/2435

A start of refactoring abstractions into `@loopback/authentication`, more stuff are in progress and feedback welcomed at any time.

### CODE REVIEW

File: packages/authentication/src/types.ts [hunk 1]
```diff
@@ -22,3 +23,31 @@ export interface UserProfile {
   name?: string;
   email?: string;
 }
+
+export type Credentials = {
+  email: string;
```
Comment (dreamdevil00) [inline packages/authentication/src/types.ts]: What if the credentials include telephone number or username?

Comment (dougal83) [reply packages/authentication/src/types.ts]: > What if the credentials include telephone number or username?

#2246 - to be addressed.

Comment (jannyHou) [reply packages/authentication/src/types.ts]: @dreamdevil00 Good point. And #2246 is supposed to provide a solution for customizing the fields. (Thanks @dougal83 )

File: packages/authentication/src/strategies/authentication-strategy.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import {inject} from '@loopback/core';
+import {AuthenticationBindings} from '../keys';
+import {AuthenticationServices} from '../services';
+import {UserProfile, AuthenticatedUser, Credentials} from '../types';
+import {Entity} from '@loopback/repository';
+import {HttpErrors} from '@loopback/rest';
+import {toJSON} from '@loopback/testlab';
```
Comment (shazha) [inline packages/authentication/src/strategies/authentication-strategy.ts]: I thought `@loopback/testlab` is supposed to be used in tests

Comment (jannyHou) [reply packages/authentication/src/strategies/authentication-strategy.ts]: Hi @shazha thank you for catching it. This PR is still in progress, code changes everyday :( I will let you know when it's ready for review. Thanks for understanding.

**Comment by jannyHou** [comment]
Closing in favour of https://github.com/strongloop/loopback-next/pull/2576.

FYI: Thank you all for the feedback again. We created a new PR that starts from adding user scenarios in docs, and follow-up PRs for creating abstractions and extensions will be coming incrementally. Let's move discussions to PR ^

**Pull request closed without merging**